[SILOptimizer] Alter FSO arg explosion heuristic.

nate-chandler opened the pull request:
The new rule on the basis of which an argument will be exploded is as follows:

(1) An argument with more than 3 live leaves after considering the results of the owned-to-guaranteed transformation will not be exploded.

(2) Rule (1) permitting, an argument all of whose non-trivial leaves are live will be exploded if and only if (a) there is more than one non-trivial leaf and (b) specialization will not result in a thunk.  This rule was beneficial before semantic ARC because it avoided collapsing RC identities.  It is not useful in the face of semantic ARC and should be retired.

(3) Rule (1) permitting, an argument with dead non-trivial leaves will be exploded.

rdar://problem/39957093

nate-chandler commented:
@swift-ci please smoke benchmark

nate-chandler commented:
@swift-ci please test

eeckstein commented on lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp:
```diff
@@ -26,42 +35,216 @@ static llvm::cl::opt<bool> FSODisableArgExplosion(
+  assert(nontrivialLeafCount >= 0 && "Counting non-trivial leaves should result in a non-negative number.");
+
+  // If the argument is completely trivial, it will not be exploded.
+  if (nontrivialLeafCount == 0) {
```
Again, this condition checks the same as the assert above.

eeckstein commented on lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp:
```diff
@@ -26,42 +35,216 @@ static llvm::cl::opt<bool> FSODisableArgExplosion(
+  // If it is known without taking the owned-to-guaranteed transformation into
+  // account both that exploding will reduce ARC traffic (because an upper bound
+  // for the number of live non-trivial leaves is less thann the non-trivial 
+  // leaf count) and also that the explosion will fit within the heuristic upper
```
Actually, when I think about it: is this really true, given that we pass most of the arguments as guaranteed by default? I think it mainly depends on the call site. Only if we need a retain-release pair at the call site, eliminating a dead argument will give a benefit.

gottesmm replied:
@eeckstein it is not that simple. What if we want to move that value pass the call site. Consider a dead guaranteed parameter. Without eliminating the argument we will not be able to reduce the lifetime and do other ARC optimizations.

I would put it this way, it only makes sense if we eliminate an ARC "constraint" on the lifetime of the non-trivially typed thing.

eeckstein reviewed (commented):
Nice!
My main concern here is that I'm not sure if exploding arguments for the sake of reducing ARC traffic helps at all, given that we pass most of our arguments as guaranteed. Some of the assumptions in the FSO come from a time where we had the owned-convention as default.
I think it's worth doing some more experiments (in case you didn't do it yet), e.g. compare benchmark results with disabling argument explosion at all, etc.

nate-chandler commented on lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp:
```diff
@@ -26,42 +35,216 @@ static llvm::cl::opt<bool> FSODisableArgExplosion(
+///    live non-trivial leaf and (b) specializing will not introduce a thunk.
+///    Breaking apart the non-trivial leaves is beneficial because it imbues
+///    the leaves with different RC identities, enabling the low level ARC 
+///    optimizer to pair retains/releases in an easier way.
```
Yup!  The comment below

```
/// TODO: Eliminate the second case.  That splitting up is only beneficial 
///       before semantic ARC.
```

is about exactly that, eliminating the second rule once we have ownership SIL at this point.  As you say, I'll run the benchmarks without this rule to see how it works.

nate-chandler replied:
Okay, opened a second PR here to run the benchmarks without this rule: https://github.com/apple/swift/pull/27252 .

nate-chandler commented:
@swift-ci please benchmark

gottesmm commented:
@nate-chandler btw the <= 6 field heuristic used by loadable from address is here in this utility: https://github.com/apple/swift/blob/master/lib/SILOptimizer/Utils/Local.cpp#L1497

nate-chandler commented:
@swift-ci please test macos

eeckstein commented on lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp:
```diff
@@ -26,42 +35,216 @@ static llvm::cl::opt<bool> FSODisableArgExplosion(
+  //
+  // Explosion makes sense only if some but not all of the leaves are live.  The
+  // dead argument transformation will try to eliminate the argument.
+  if (liveLeafCountUpperBound == 0) {
```
Forget this comment, I didn't see that the assert is >= 0 and not > 0

gottesmm commented on include/swift/SIL/Projection.h:
```diff
@@ -751,13 +752,15 @@ class ProjectionTreeNode {
+  Optional<Projection> &getProjection() { return Proj; }
 
-  llvm::SmallVector<Operand *, 4> getNonProjUsers() const {
+  const SmallVectorImpl<Operand *> &getNonProjUsers() const {
```
As a quick aside: If this is a const, better to return an ArrayRef.

nate-chandler replied:
Fixed.

Catfish-Man commented:
I'm slightly concerned the positive impact of this might be overstated a bit in the benchmark results, since it looks like most of the wins are just repeated variations of 2 tests.

gottesmm commented on lib/SILOptimizer/FunctionSignatureTransforms/FunctionSignatureOpts.h:
```diff
@@ -214,6 +225,18 @@ struct FunctionSignatureTransformDescriptor {
+  /// specialization.
+  bool willSpecializationIntroduceThunk() {
+    return !hasOnlyDirectInModuleCallers ||
+           hasPublicVisibility(OriginalFunction->getLinkage());
```
I am not sure if hasPublicVisibility is the right check here. Is that a c function? If so, does this properly handle WMO vs non-WMO?

nate-chandler replied:
Fixed.

nate-chandler commented:
@swift-ci please test compiler performance

gottesmm reviewed (commented):
Quick comments

nate-chandler commented:
@Catfish-Man That's fair.  Still doing measurements over here and specifically will be looking at the code size impact on more realistic targets including the compatibility suite.

nate-chandler committed:
[SILOptimizer] ProjectionTree can vend leaf types.

Added getAllLeafTypes to ProjectionTree.  The new method vends, via an
out paramter, a vector containing the types of all the leaves in a
projection tree in the order that they appear.  The method relies uses a
new convenience on ProjectionTreeNode, isLeaf to include only the types
of those nodes which are leaves.

Excerpted from @gottesm's https://github.com/apple/swift/pull/16756.

nate-chandler committed:
[SILOptimizer] Projection tree can vend users.

Added getUsers to ProjectionTree.  The new method vands, via an out
parameter, a set of all the users of all of the nodes in the projection
tree that are themselves not in the projection tree by way of
getNonProjUsers.  Took this opportunity to tweak getNonProjUsers to vend
a const ArrayRef rather than a SmallVector.

Excerpted from @gottesm's https://github.com/apple/swift/pull/16756.

nate-chandler committed:
[SILOptimizer] Epilogue matcher can find partial post dom argument releases.

Added getPartiallyPostDomReleaseSet to
ConsumedArgToEpilogueReleaseMatcher.  Given an argument, the new method
returns the array of releases of the argument if there is an array
thereof and if the releases therein do not jointly post-dominate the
argument.

Excerpted from @gottesm's https://github.com/apple/swift/pull/16756.

nate-chandler committed:
[Gardening] Minor cleanup to Projection.h.

Replaced some namespace qualified references to ArrayRef and Optional
with the unqualified type.  Reordered the includes per clang-format.

Excerpted from @gottesm's https://github.com/apple/swift/pull/16756.

nate-chandler committed:
[Gardening] Documented ArgumentDescriptor method.

Added brief doc for FunctionSignatureOpts' ArgumentDescriptor's method
canOptimizeLiveArg and tweaked the style to add braces around the body
of a single line if clause.

nate-chandler committed:
[Gardening] Made FSOEnableGenerics option static.

The command-line option for sil-fso-enable-generics was previously
visible outside the FunctionSignatureOpts translation unit.  It is not
any longer.

nate-chandler committed:
[SILOptimizer] Added option to force FSO for uncalled functions.

The new flag -sil-fso-optimize-if-not-called forced function signature
optimization to run even on functions which are not called.  Doing so is
helpful for tests to alleviate the burden of writing code to actually
call a function in whose function signature optimization we are
interested.

nate-chandler commented:
```
LLVM.NumLLVMBytesOutput | 1,797,132,560 | 1,797,052,058 | -80,502 | -0.0%
```

from https://ci.swift.org/job/swift-PR-compiler-performance-macOS/539/artifact/comment.md

nate-chandler commented:
```
LLVM.NumLLVMBytesOutput | 1,535,314,418 | 1,528,530,034 | -6,784,384 | -0.44%
```

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
The heuristic is now to explode arguments if one of the following sets of conditions hold:

(1) (a) Specializing the function will result in a thunk.  That is, the
        thunk that is generated cannot be inlined everywhere.
    (b) The argument has dead non-trivial leaves.
    (c) The argument has fewer than three live leaves.

(2) (a) Specializing the function will not result in a thunk.  That is,
        the thunk that is generated will be inlined everywhere and
        eliminated as dead code.
    (b) The argument has dead potentially trivial leaves.
    (c) The argument has fewer than six live leaves.

nate-chandler commented:
@swift-ci please test linux

nate-chandler commented:
@swift-ci please test compiler performance

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
Debug:
```
LLVM.NumLLVMBytesOutput | 1,797,952,710 | 1,797,869,638 | -83,072 | -0.0%
```

Release:
```
LLVM.NumLLVMBytesOutput | 1,921,660,574 | 1,911,421,822 | -10,238,752 | -0.53%
```

From https://ci.swift.org/job/swift-PR-compiler-performance-macOS/541/artifact/comment.md of https://ci.swift.org/job/swift-PR-compiler-performance-macOS/541/

nate-chandler commented:
@swift-ci please test macos platform

nate-chandler commented on lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp:
```diff
@@ -26,42 +35,216 @@ static llvm::cl::opt<bool> FSODisableArgExplosion(
+///
+/// 2. Values that are completely trivial. By splitting these up we create
+///    more register pressure during argument marshalling and do not really add
+///    any advantage. We only eliminate them
```
Fixed!

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
[CallerAnalysis] foundAllCallers reqs nonexternal.

Previously, CallerAnalysis::FunctionInfo.foundAllCallers(), which is
documented to return true only when specialization of a function will
not require a thunk, returned true for functions which are possibly used
externally.  Now, that member function only returns false for functions
which may be used externally since dead code elimination will not be
able to remove them.

nate-chandler committed:
[SILOptimizer] Alter FSO arg explosion heuristic.

The new rule is that an argument will be exploded if one of the
following sets of conditions hold:

(1) (a) Specializing the function will result in a thunk.  That is, the
        thunk that is generated cannot be inlined everywhere.
    (b) The argument has dead non-trivial leaves.
    (c) The argument has fewer than three live leaves.

(2) (a) Specializing the function will not result in a thunk.  That is,
        the thunk that is generated will be inlined everywhere and
        eliminated as dead code.
    (b) The argument has dead potentially trivial leaves.
    (c) The argument has fewer than six live leaves.

This change is based heavily on @gottesm's
https://github.com/apple/swift/pull/16756 .

rdar://problem/39957093

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test compiler performance

nate-chandler commented:
@swift-ci please test macos platform

nate-chandler commented:
@swift-ci please clean smoke test macos platform

nate-chandler commented:
Release:

```
LLVM.NumLLVMBytesOutput | 1,924,598,716 | 1,920,184,346 | -4,414,370 | -0.23%
```

Debug:

```
LLVM.NumLLVMBytesOutput | 1,796,502,774 | 1,796,441,004 | -61,770 | -0.0%
```

Source:

https://ci.swift.org/job/swift-PR-compiler-performance-macOS/544/artifact/

nate-chandler commented:
@swift-ci please clean test macos platform

nate-chandler commented:
@swift-ci please clean test macos platform
